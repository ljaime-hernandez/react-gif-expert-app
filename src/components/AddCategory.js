import React, {useState} from 'react'
import PropTypes from 'prop-types';

// the props 'setCategories' are destructured before being used on the component, done that we can call the 
// setCategories function from the parent component to properly add the input value into the string array declared
// on it
export const AddCategory = ({setCategories}) => {
    
    // the inputValue by default wll be left as an empty string on the useState, if we remove the 
    // quotation marks then the browser will send an error, the error will not be explicit but it will be
    // a reference to the inputValue object stating that it is in an undefined state, when the 
    // quotation marks are placed on the useState then the error will disappear
    const [inputValue, setInputValue] = useState('');

    // this function will retrieve the value placed in the input box, done by calling the event, 
    // then the target value will contain the input value which will be assigned to the inputValue
    // object which we will use to include into the categories from the previous component
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    // we use the handleSubmit function to first prevent the input box from having an unrequired submission
    // with an incorrect input, then we check if the input value is greater than 2 characters, if its true
    // then the prop setCategories from the previous component is called for it to include the string input in
    // the categories object
    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2){
            // inputvalue is placed first for it to push the other results to the bottom on the screen and have
            // the new results for quick reading
            setCategories(category => [inputValue, ...category]);
            // sets the text input box to an empty state
            setInputValue('');
        }
        
        console.log('Submit succeed')
    }

    return (
        <form onSubmit={handleSubmit}>
            {/*'p' element added just for testing purposes, more information on the file located in
            tests/components/AddCategory.test.js file*/}
            <p hidden>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

// the PropTypes will send an error on the browser console if the AddCategory component is used
// without the setCategories sent as an argument, this will add a security layer which will allow
// to only make modifications to a get or post request under our rules
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}