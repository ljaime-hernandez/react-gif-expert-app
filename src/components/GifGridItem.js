import React from 'react';

export const GifGridItem = ({title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
 