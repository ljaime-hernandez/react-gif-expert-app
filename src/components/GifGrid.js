import React from 'react'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({category}) => {

    // const [images, setImages] = useState([]);

    const {data:images, loading} = useFetchGifs(category);

    // // function run every time the component is loaded
    // useEffect(() => {
    //     getGifs(category)
    //         .then(imgs => setImages(imgs));
    //         // also .then(setImages);

    // // in determined cases, the argument sent from the parent component might change instead of adding an additional
    // // one, a warning will be sent through the console for us to acknowledge a lack of cohesion on the useEffect, in
    // // this code we just need to add the category argument as a dependency inside the array at the end of the function     
    // }, [category])

    return (
        <>
            <h3>{category}</h3>

            {loading && <p>Loading</p>}
            {/* the operator && will work for us as an evaluation of a true/false statement, it can be used
            instead of a ternary evaluation, the ternary option is placed below as example for a different approach.
            {loading ? <p>Loading</p> : null}
            */}

            {/*we use className instead of class as javascript will give us a warning when rendering the tag, as "class"
            is a reserved word for scripting uses*/}
             <div className="card-grid">
                {
                    
                    images.map( image => (
                            <GifGridItem 
                                key={image.id}
                                {...image} />))
                }
            </div> 
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}