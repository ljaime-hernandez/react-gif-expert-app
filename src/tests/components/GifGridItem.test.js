import React from "react";
import { shallow } from "enzyme";
import { GifGridItem } from "../../components/GifGridItem"

/*
we are using the shallow method of enzyme for us to test the results expected from a component.
looking back on the setupTests.js file, we can confirm two imports which would help us to understand
the tests appropriately. First, there is an adapter used for enzyme to work on react 17, which is the
actual version used in this code, and as second requirement, we are using an import from a library
called "enzyme to json", which works as a serializer for enzyme, what this does is that it is going to
help us on the creation of the enzyme snapshot used on the component to test the component with various
properties.
*/

describe('Tests on <GifGridItem />', () =>{

    const title = 'Pokemon';
    const url = 'https://pokemon.org';
    // the shallow function here (along with the enzyme to json serializer) will create a snapshot of the
    //component by taking it as a reference along with whatever properties we pass on it to be tested
    const wrapper = shallow(<GifGridItem title={title} url={url}/>);

    test('should display component correctly', () => {
        
        expect(wrapper).toMatchSnapshot();
    })

    test('should have a paragraph on the title', () => {
        
        const p = wrapper.find('p');
        expect(p.text().trim()).toBe(title);
    })
    
    test('should have a picture similar to the url and alt of the properties', () => {
        
        // the find method is used in here to find the img element, which in this case will
        //contain the src and the alt properties of the image. if we use a console.log and use the
        // props() method on it, the console will return an object with the properties contained on it.
        // we can also use the prop() method without the "s" on it, and put an specific property as argument
        // for it to be returned as necessary, 
        // ex: img.prop('src');      ----------------->   https://pokemon.org
        const img = wrapper.find('img');
        // console.log(img.prop('src'));
        expect(img.prop('src')).toBe(url);
        expect(img.prop('alt')).toBe(title);
    })

    test('should have class animate__fadeInLeft', () => {
        
        const divClass = wrapper.find('div');
        const divClassText = divClass.prop('className');
        expect(divClassText).toContain('animate__fadeInLeft');
        // also expect(divClassText.includes('animate__fadeInLeft')).toBe(true);
        // also expect(divClassText.includes('animate__fadeInLeft')).not.toBe(false);
    })
})
