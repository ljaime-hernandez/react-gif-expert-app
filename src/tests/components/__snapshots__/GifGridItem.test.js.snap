// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <GifGridItem /> should display component correctly 1`] = `
<div
  className="card animate__animated animate__fadeInLeft"
>
  <img
    alt="Pokemon"
    src="https://pokemon.org"
  />
  <p>
    Pokemon
  </p>
</div>
`;
